# Travaux Pratiques



Data Science - Règles d’Association
décembre 2023

---



## **1. Chargez les données Groceries.**

In [ ]:
install.packages("arules")


Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



(a) Appliquez l’algorithme apriori pour extraire des règles d’association

i. extrayez toutes les RA.

In [ ]:
library(arules)

# Lire les données (s'assurer que le chemin du fichier est correct)
transactions <- read.transactions('store_data.csv', format = 'basket')

# Application de l'algorithme Apriori
rules <- apriori(transactions, parameter = list(supp = 0.01, conf = 0.1))

# Afficher les règles
inspect(rules)


Loading required package: Matrix


Attaching package: ‘arules’


The following objects are masked from ‘package:base’:

    abbreviate, write


Warning message in asMethod(object):
“removing duplicated items in transactions”


Apriori

Parameter specification:
 confidence minval smax arem  aval originalSupport maxtime support minlen
        0.1    0.1    1 none FALSE            TRUE       5    0.01      1
 maxlen target  ext
     10  rules TRUE

Algorithmic control:
 filter tree heap memopt load sort verbose
    0.1 TRUE TRUE  FALSE TRUE    2    TRUE

Absolute minimum support count: 75 

set item appearances ...[0 item(s)] done [0.00s].
set transactions ...[5729 item(s), 7501 transaction(s)] done [0.01s].
sorting and recoding items ... [45 item(s)] done [0.00s].
creating transaction tree ... done [0.00s].
checking subsets of size 1 2 done [0.00s].
writing ... [19 rule(s)] done [0.00s].
creating S4 object  ... done [0.00s].
     lhs          rhs       support    confidence coverage   lift      count
[1]  {}        => {tea}     0.10705239 0.1070524  1.00000000  1.000000 803  
[2]  {rice}    => {wheat}   0.01226503 0.9583333  0.01279829 11.144897  92  
[3]  {wheat}   => {rice}    0.01226503 0.1426357  0.0859885

**Interprétation des Résultats:**

Support: Le support d'une règle indique la fréquence à laquelle les items de la règle apparaissent ensemble dans l'ensemble de données. Par exemple, la règle {tea} => {green} a un support de 0.01186508, ce qui signifie que 'tea' et 'green' apparaissent ensemble dans environ 1.19% de toutes les transactions.
Confiance: La confiance mesure la fiabilité de l'inférence faite par une règle. Par exemple, la règle {sauce} => {cream} a une confiance de 0.7938144, ce qui signifie que dans 79.38% des cas où 'sauce' est achetée, 'cream' l'est aussi.
Lift: Le lift mesure à quel point la présence d'un item (antécédent) augmente la probabilité de voir l'autre item (conséquent). Un lift supérieur à 1 indique une association positive. Par exemple, la règle {herb} => {&} a un lift de 20.218329, ce qui est significativement élevé, indiquant une forte association entre ces deux items.

**Observations Notables:**
Plusieurs règles ont un lift très élevé, comme {herb} => {&} et {sauce} => {cream}, ce qui suggère des associations très fortes.
Certaines règles, comme {rice} => {wheat} et {whole} => {wheat}, ont à la fois une confiance et un lift élevés, indiquant qu'elles sont à la fois fiables et significatives.

**Conclusion:**
Ces règles d'association peuvent être utiles pour comprendre les comportements d'achat et pour les stratégies de marketing croisé. Par exemple, si un client achète 'sauce', il est très probable qu'il achète aussi 'cream'.
Cependant, il est important de tenir compte de la couverture (fréquence relative des antécédents) pour s'assurer que les règles sont applicables à un nombre suffisant de cas.

ii. imposez selon votre fantaisie une partie droite, puis une partie gauche.

In [ ]:
# Imposer une partie droite (consequents) avec l'item 'milk'
rules_with_milk <- subset(rules, rhs %in% "milk")

# Imposer une partie gauche (antecedents) avec l'item 'bread'
rules_with_bread <- subset(rules, lhs %in% "bread")

# Afficher les règles
inspect(rules_with_milk)
inspect(rules_with_bread)



/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


(b) Appliquez l’algorithme apriori pour extraire des itemsets fréquents et maximaux.

In [ ]:

# Appliquer l'algorithme Apriori pour trouver les itemsets fréquents
frequent_itemsets <- apriori(transactions, parameter = list(supp = 0.01, target = "frequent itemsets"))

# Afficher les itemsets fréquents
inspect(frequent_itemsets)

# Appliquer l'algorithme Apriori pour trouver les itemsets maximaux
maximal_itemsets <- apriori(transactions, parameter = list(supp = 0.01, target = "maximal itemsets"))

# Afficher les itemsets maximaux
inspect(maximal_itemsets)


Apriori

Parameter specification:
 confidence minval smax arem  aval originalSupport maxtime support minlen
         NA    0.1    1 none FALSE            TRUE       5    0.01      1
 maxlen            target  ext
     10 frequent itemsets TRUE

Algorithmic control:
 filter tree heap memopt load sort verbose
    0.1 TRUE TRUE  FALSE TRUE    2    TRUE

Absolute minimum support count: 75 

set item appearances ...[0 item(s)] done [0.00s].
set transactions ...[5729 item(s), 7501 transaction(s)] done [0.02s].
sorting and recoding items ... [45 item(s)] done [0.00s].
creating transaction tree ... done [0.00s].
checking subsets of size 1 2 done [0.00s].
sorting transactions ... done [0.00s].
writing ... [54 set(s)] done [0.00s].
creating S4 object  ... done [0.00s].
     items                    support    count
[1]  {eggs}                   0.01199840  90  
[2]  {escalope}               0.01346487 101  
[3]  {chocolate}              0.01359819 102  
[4]  {energy}                 0.01133182  

ERROR: ignored

**Interprétation des Résultats :**
Chaque ligne dans vos résultats représente un itemset fréquent, avec le support indiqué pour chaque itemset. Par exemple, l'itemset {eggs} apparaît dans environ 1.2% de toutes les transactions (support = 0.01199840).
Les itemsets sont variés, allant des produits individuels (comme {eggs} ou {escalope}) à des combinaisons de produits (comme {shrimp, frozen} ou {spaghetti, mineral}).

**Itemsets Notables :**
Certains itemsets comme {tea} et {wheat} apparaissent fréquemment dans les transactions, avec des supports respectifs de 0.10705239 et 0.08598853, ce qui indique leur popularité.
Les combinaisons d'items comme {rice, wheat} et {cream, sauce} montrent des associations spécifiques entre produits. Ces associations peuvent être utiles pour des stratégies de marketing croisé ou pour la gestion des stocks.

**Analyse des Supports :**
Les supports varient considérablement, allant de 1.2% pour des items individuels à plus de 10% pour des items très courants comme {tea}.
Des supports plus élevés suggèrent des items ou des combinaisons d'items particulièrement populaires ou fréquemment achetés ensemble.

(c) Appliquez l’algorithme eclat dans le même but.

In [ ]:
# Appliquer l'algorithme Eclat pour trouver les itemsets fréquents
frequent_itemsets_eclat <- eclat(transactions, parameter = list(supp = 0.01, maxlen = 10))

# Afficher les itemsets fréquents
inspect(frequent_itemsets_eclat)

# Trouver les itemsets maximaux parmi les itemsets fréquents
maximal_itemsets_eclat <- which(colSums(is.subset(frequent_itemsets_eclat, frequent_itemsets_eclat)) == 1)
maximal_itemsets_eclat <- frequent_itemsets_eclat[maximal_itemsets_eclat]

# Afficher les itemsets maximaux
inspect(maximal_itemsets_eclat)



Eclat

parameter specification:
 tidLists support minlen maxlen            target  ext
    FALSE    0.01      1     10 frequent itemsets TRUE

algorithmic control:
 sparse sort verbose
      7   -2    TRUE

Absolute minimum support count: 75 

create itemset ... 
set transactions ...[5729 item(s), 7501 transaction(s)] done [0.01s].
sorting and recoding items ... [45 item(s)] done [0.00s].
creating sparse bit matrix ... [45 row(s), 7501 column(s)] done [0.00s].
writing  ... [54 set(s)] done [0.00s].
Creating S4 object  ... done [0.00s].
     items                    support    count
[1]  {rice, wheat}            0.01226503  92  
[2]  {cream, sauce}           0.01026530  77  
[3]  {mineral, water}         0.01239835  93  
[4]  {green, tea}             0.01186508  89  
[5]  {french, fries}          0.01786428 134  
[6]  {wheat, whole}           0.01893081 142  
[7]  {&, herb}                0.03092921 232  
[8]  {fat, yogurt}            0.06839088 513  
[9]  {tea, wheat}             0.011

**Interprétation des Résultats :**
Les itemsets listés sont ceux qui apparaissent fréquemment dans votre base de données. Chaque itemset est accompagné de son support et du nombre de fois qu'il apparaît (count).
Par exemple, l'itemset {tea} a un support de 0.10705239, signifiant qu'il apparaît dans environ 10.7% de toutes les transactions, et il apparaît 803 fois dans l'ensemble des données.

**Itemsets Individuels vs. Combinés :**
Vos résultats montrent à la fois des itemsets individuels (comme {wheat}, {tea}) et des combinaisons d'items (comme {rice, wheat}, {cream, sauce}).
Les itemsets individuels avec des supports élevés comme {tea} et {wheat} indiquent leur popularité générale.
Les combinaisons d'items révèlent des achats fréquemment réalisés ensemble, ce qui peut être utile pour les recommandations de vente croisée ou pour comprendre les habitudes d'achat.

(d) Utilisez la fonction system.time pour comparer le temps des deux
algorithmes.

In [ ]:
# Mesurer le temps d'exécution de l'algorithme Apriori
start_time <- Sys.time()
rules <- apriori(transactions, parameter = list(supp = 0.01, conf = 0.1))
end_time <- Sys.time()
execution_time <- end_time - start_time

# Afficher le temps d'exécution
print(execution_time)

# Mesurer le temps d'exécution pour Eclat
start_time_eclat <- Sys.time()
itemsets_eclat <- eclat(transactions, parameter = list(supp = 0.01, maxlen = 10))
end_time_eclat <- Sys.time()
time_eclat <- end_time_eclat - start_time_eclat

# Afficher le temps d'exécution pour Eclat
print(time_eclat)



Apriori

Parameter specification:
 confidence minval smax arem  aval originalSupport maxtime support minlen
        0.1    0.1    1 none FALSE            TRUE       5    0.01      1
 maxlen target  ext
     10  rules TRUE

Algorithmic control:
 filter tree heap memopt load sort verbose
    0.1 TRUE TRUE  FALSE TRUE    2    TRUE

Absolute minimum support count: 75 

set item appearances ...[0 item(s)] done [0.00s].
set transactions ...[5729 item(s), 7501 transaction(s)] done [0.02s].
sorting and recoding items ... [45 item(s)] done [0.00s].
creating transaction tree ... done [0.00s].
checking subsets of size 1 2 done [0.00s].
writing ... [19 rule(s)] done [0.00s].
creating S4 object  ... done [0.00s].
Time difference of 0.06727147 secs
Eclat

parameter specification:
 tidLists support minlen maxlen            target  ext
    FALSE    0.01      1     10 frequent itemsets TRUE

algorithmic control:
 sparse sort verbose
      7   -2    TRUE

Absolute minimum support count: 75 

create item

**Analyse des Performances de l'Algorithme Apriori :**
* Temps d'Exécution : L'algorithme Apriori a pris environ 0.067 secondes (Time difference of 0.06727147 secs) pour exécuter. Ce temps est relativement rapide, montrant une bonne efficacité pour la taille de votre ensemble de données.

* Nombre de Règles Générées : Apriori a généré 19 règles d'association, ce qui indique que peu de règles répondent aux critères de support et de confiance définis (support minimum de 0.01 et confiance minimum de 0.1).

* Utilité pour l'Analyse : Ces règles peuvent offrir des insights pertinents sur les relations entre différents items dans les transactions. Elles sont utiles pour comprendre les habitudes d'achat des clients.

**Analyse des Performances de l'Algorithme Eclat :**

* Temps d'Exécution : Eclat a pris environ 0.045 secondes (Time difference of 0.04491425 secs). Cela suggère que Eclat est légèrement plus rapide qu'Apriori pour ce jeu de données.

* Nombre d'Itemsets Fréquents Générés : Eclat a généré 54 itemsets fréquents. Cela montre que plus d'associations répondent au critère de support minimum par rapport aux règles d'association générées par Apriori.

* Utilité pour l'Analyse : Les itemsets fréquents identifiés par Eclat peuvent aider à comprendre quels groupes d'items sont souvent achetés ensemble, ce qui est utile pour la planification des stocks et les promotions.

**Comparaison :**

* Efficacité : Eclat s'est avéré légèrement plus rapide qu'Apriori dans ce cas, ce qui est cohérent avec les attentes, car Eclat est généralement plus efficace pour de grands ensembles de données.

* Résultats : Alors qu'Apriori se concentre sur les règles d'association, Eclat identifie les itemsets fréquents. Les deux types de résultats sont utiles mais servent des objectifs analytiques légèrement différents.

* Choix de l'Algorithme : Le choix entre Apriori et Eclat dépendra de l'objectif spécifique de votre analyse. Pour des règles d'association spécifiques, Apriori est préférable. Pour identifier rapidement des itemsets fréquents, Eclat est plus adapté.

* Considérations sur les Paramètres : Les paramètres tels que le support minimum et la confiance jouent un rôle crucial dans le nombre de règles ou d'itemsets générés. Ils doivent être soigneusement ajustés en fonction des besoins de l'analyse.

(e) Construisez une collection BigGr beaucoup plus grande de transac- tions issues de Groceries (avec sample(), par exemple).

In [ ]:

# Dupliquer les transactions pour créer BigGr
set.seed(123)  # Pour la reproductibilité
biggr_transactions <- sample(transactions, size = length(transactions) * 5, replace = TRUE)


(f) Appliquez à nouveau apriori et eclat pour les itemsets fréquents et fermés ("closed frequent itemset") et comparez à nouveau les temps de calcul.

In [ ]:
# Appliquer Apriori pour les itemsets fréquents sur BigGr
start_time_apriori_biggr <- Sys.time()
frequent_itemsets_apriori_biggr <- apriori(biggr_transactions, parameter = list(supp = 0.01, target = "frequent itemsets"))
end_time_apriori_biggr <- Sys.time()
time_apriori_biggr <- end_time_apriori_biggr - start_time_apriori_biggr

# Afficher le temps d'exécution pour Apriori sur BigGr
print(paste("Temps d'exécution pour Apriori sur BigGr:", time_apriori_biggr))




# Appliquer Eclat pour les itemsets fréquents sur BigGr
start_time_eclat_biggr <- Sys.time()
frequent_itemsets_eclat_biggr <- eclat(biggr_transactions, parameter = list(supp = 0.01, maxlen = 10))
end_time_eclat_biggr <- Sys.time()
time_eclat_biggr <- end_time_eclat_biggr - start_time_eclat_biggr

# Afficher le temps d'exécution pour Eclat sur BigGr
print(paste("Temps d'exécution pour Eclat sur BigGr:", time_eclat_biggr))


Apriori

Parameter specification:
 confidence minval smax arem  aval originalSupport maxtime support minlen
         NA    0.1    1 none FALSE            TRUE       5    0.01      1
 maxlen            target  ext
     10 frequent itemsets TRUE

Algorithmic control:
 filter tree heap memopt load sort verbose
    0.1 TRUE TRUE  FALSE TRUE    2    TRUE

Absolute minimum support count: 375 

set item appearances ...[0 item(s)] done [0.00s].
set transactions ...[5694 item(s), 37505 transaction(s)] done [0.05s].
sorting and recoding items ... [46 item(s)] done [0.00s].
creating transaction tree ... done [0.01s].
checking subsets of size 1 2 done [0.00s].
sorting transactions ... done [0.01s].
writing ... [54 set(s)] done [0.00s].
creating S4 object  ... done [0.00s].
[1] "Temps d'exécution pour Apriori sur BigGr: 0.0947248935699463"
Eclat

parameter specification:
 tidLists support minlen maxlen            target  ext
    FALSE    0.01      1     10 frequent itemsets TRUE

algorithmic contro

**Analyse de l'Algorithme Apriori sur BigGr :**

* Paramètres de l'Algorithme : Le support minimum est fixé à 0.01, et l'objectif est de trouver des itemsets fréquents. La taille maximale des itemsets est de 10.

* Nombre d'Itemsets Fréquents Générés : Apriori a identifié 54 itemsets fréquents dans une collection de 37,505 transactions, ce qui indique une bonne identification des patterns fréquents dans un ensemble de données plus grand.

* Temps d'Exécution : Le temps d'exécution pour Apriori sur BigGr est d'environ 0.095 secondes. Cette performance montre que l'algorithme gère bien l'augmentation de la taille des données.

**Analyse de l'Algorithme Eclat sur BigGr :**
* Paramètres de l'Algorithme : Eclat a également été exécuté avec un support minimum de 0.01 pour identifier les itemsets fréquents.

* Nombre d'Itemsets Fréquents Générés : Eclat a également généré 54 itemsets fréquents, ce qui est cohérent avec les résultats d'Apriori.

* Temps d'Exécution : Le temps d'exécution pour Eclat sur BigGr est d'environ 0.094 secondes, légèrement plus rapide que Apriori, mais la différence est minime.

**Comparaison et Conclusion :**

* Performance des Deux Algorithmes : Les deux algorithmes ont montré une efficacité similaire en termes de temps d'exécution sur un grand ensemble de données, avec Eclat légèrement plus rapide que Apriori.

* Cohérence des Résultats : Le fait que les deux algorithmes aient identifié le même nombre d'itemsets fréquents renforce la fiabilité des résultats.

* Implications pour le TP : Ces résultats démontrent que les deux algorithmes sont bien adaptés pour traiter des ensembles de données volumineux, et ils fournissent des insights précieux sur les tendances d'achat dans un ensemble de données étendu.

##**2. (optionnel) Réalisez sur la grande collection BigGr une implémentation de Partition-apriori**.

Il y a la fonction duplicated qui détecte des "éléments" (itemsets, tran- sactions ou règles d’association) redondants.

In [ ]:

# Diviser les transactions en partitions (par exemple, en 5 partitions)
partition_size <- length(biggr_transactions) / 5
partitions <- split(biggr_transactions, ceiling(seq_along(biggr_transactions) / partition_size))

# Fonction pour appliquer Apriori à une partition
apply_apriori <- function(partition, min_support = 0.01) {
    return(apriori(partition, parameter = list(supp = min_support, target = "frequent itemsets")))
}

# Appliquer Apriori à chaque partition et stocker les résultats
results <- lapply(partitions, apply_apriori)


# Fonction pour trouver les itemsets communs
find_common_itemsets <- function(results) {
    common_itemsets <- results[[1]]
    for (result in results[-1]) {
        common_itemsets <- intersect(common_itemsets, result)
    }
    return(common_itemsets)
}

# Trouver les itemsets communs
common_itemsets <- find_common_itemsets(results)

# Éliminer les itemsets redondants
unique_itemsets <- unique(common_itemsets)


Apriori

Parameter specification:
 confidence minval smax arem  aval originalSupport maxtime support minlen
         NA    0.1    1 none FALSE            TRUE       5    0.01      1
 maxlen            target  ext
     10 frequent itemsets TRUE

Algorithmic control:
 filter tree heap memopt load sort verbose
    0.1 TRUE TRUE  FALSE TRUE    2    TRUE

Absolute minimum support count: 75 

set item appearances ...[0 item(s)] done [0.00s].
set transactions ...[4128 item(s), 7501 transaction(s)] done [0.02s].
sorting and recoding items ... [47 item(s)] done [0.00s].
creating transaction tree ... done [0.00s].
checking subsets of size 1 2 done [0.00s].
sorting transactions ... done [0.00s].
writing ... [55 set(s)] done [0.00s].
creating S4 object  ... done [0.00s].
Apriori

Parameter specification:
 confidence minval smax arem  aval originalSupport maxtime support minlen
         NA    0.1    1 none FALSE            TRUE       5    0.01      1
 maxlen            target  ext
     10 frequent 

**Analyse des Résultats de Partition-Apriori :**

 **Nombre d'Itemsets Fréquents Identifiés :**
 * Chaque exécution de l'algorithme Apriori sur les différentes partitions a identifié un nombre variable d'itemsets fréquents, allant de 48 à 56 itemsets.

* Cette variation indique que différents segments de votre ensemble de données (représentés par chaque partition) possèdent des caractéristiques et des associations d'items uniques.

**Variabilité des Itemsets :**
* Le nombre d'items uniques dans chaque partition varie, allant de 3961 à 4128, ce qui peut influencer le nombre d'itemsets fréquents identifiés.

* Cette variation souligne l'importance de la diversité des transactions au sein de chaque partition et son impact sur l'identification des patterns fréquents.


**Implications pour l'Analyse de Données :**
* Ces résultats montrent que l'utilisation de l'algorithme Partition-Apriori permet de capturer des nuances dans les associations d'items qui pourraient être manquées en analysant l'ensemble des données en bloc.

* Les différences observées entre les partitions pourraient fournir des insights sur les variations de comportement d'achat ou des préférences au sein de sous-groupes de votre base de clients.




##**Conclusion**
L'approche Partition-Apriori révèle des informations détaillées et spécifiques à certaines parties de l'ensemble des données, ce qui peut être particulièrement utile pour une analyse ciblée ou segmentée.

Les variations dans les itemsets fréquents identifiés soulignent l'importance de considérer la diversité des transactions lors de l'analyse des habitudes d'achat.

Ces résultats pourraient être utilisés pour formuler des recommandations spécifiques à certaines catégories de produits ou pour améliorer les stratégies de marketing ciblé.